pkg/dynamic_pool.go

```diff
@@ -0,0 +1,90 @@
+package datago
+
+import (
+	"fmt"
+	"runtime"
+	"time"
+)
+
+// Define an enum which will be used to track the state of the worker
+type worker_state int
+
+const (
+	worker_idle worker_state = iota
+	worker_running
+	worker_done
+	worker_stopping
+)
+
+// Define a stateful worker struct which will be spawned by the worker pool
+type worker struct {
+	state worker_state
+}
+
+// Manage a pool of workers to fetch the samples
+// We'll initially spawn half the machine capacity in terms of workers,
+// and then we'll dynamically adjust the number of workers based on the
+// work backlog and idle time
+
+func run_worker_pool(sampleWorker func(*worker), chanInputs *BufferedChan[SampleDataPointers], chanOutputs *BufferedChan[Sample]) {
+
+	// Get the number of CPUs on the machine
+	max_workers := runtime.NumCPU() // We suppose that this doesn´t change during the execution
+	min_workers := int32(max_workers / 8)
+	worker_pool_size := max_workers / 2
+	const response_time = 1 * time.Second
+	fmt.Println("Worker pool started. Initial size: ", worker_pool_size, " min_workers: ", min_workers, " max_workers: ", max_workers)
+
+	// Start the workers and work on the metadata channel
+	var workers []*worker
+
+	for i := 0; i < worker_pool_size; i++ {
+		new_worker := worker{state: worker_idle}
+		workers = append(workers, &new_worker)
+		go sampleWorker(&new_worker)
+	}
+
+	// Every second, check the state of the workers and adjust the pool size
+	// based on the work backlog and idle time
+	for {
+		// Create new worker
```

Comment:
the logic is there: if there's a significant backlog & free space on the machine, grow the pool. If too many workers are idle, shrink the pool